subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/caching/CachedMissingModulesIntegrationTest.groovy

```diff
@@ -193,6 +235,137 @@ Required by:
         run 'retrieve'
     }
 
+    def "cached empty version list is ignored when looking for different version range"() {
```

Comment:
The name of this test is not right: it's not because we are looking for a different version range that we ignore the cached empty version list. It's because if we trusted the cache, we'd fail to resolve altogether.

The behaviour demonstrated here is that Gradle first tries to resolve trusting it's caches, but before failing will make a last-ditch effort, assuming that what's in the cache might be stale.

The primary use case for this behaviour is:
1. User thinks they published something, but they messed it up
2. User tries to resolve thing, and resolve fails. We cache the fact that thing is missing.
3. User publishes thing correctly
4. User resolves again: they don't need to use `--refresh-dependencies` for Gradle to succeed.